README.md
# Format Preserving Encryption in Java

An implementation of the NIST-approved FF1 and FF3-1 algorithms in Java.

This implementation conforms (as best as possible) to
[Draft SP 800-38G Rev. 1][800-38g1]. The implementation passes all tests
specified by NIST in their Cryptographic Standards and Guidelines
[examples for FF1][ff1-examples]; however, no official examples/samples exist
(or are known) for FF3-1. FF3 is not implemented as NIST has officially
deprecated its use in light of recent [cryptanalysis][ff3-cryptanalysis]
performed on it.

# Building

The library is dependent on [Bouncy Castle](https://www.bouncycastle.org/)
and [JUnit](https://junit.org/junit4/). In addition, you'll need to have
[gradle](https://gradle.org/) installed to do the build:
```sh
$ ./gradlew build
```
The above commands will build the library.

# Testing

To run the tests:
```sh
$ ./gradlew test
```
or to force (re)running the tests:
```sh
$ ./gradlew cleanTest test
```
As described above, the unit tests for FF1 come from the NIST guidelines. As
no such guidelines are available for FF3-1, the unit tests verify only that
the encryption and decryption implementations are compatible with each other.

# Documentation

The interfaces are documented in the source
[files](lib/src/main/java/ubiqsecurity/fpe).

Additionally, documentation can be produced via gradle:
```sh
$ ./gradlew javadoc
```
which will produce HTML documentation in `lib/build/docs/javadoc`.

### About alphabets and the radix parameter

The interfaces operate on strings, and the radix parameter determines which
characters are valid within those strings, i.e. the alphabet. For example, if
your radix is 10, then the alphabet for your plain text consists of the
characters in the string "0123456789". If your radix is 16, then the
alphabet is the characters in the string "0123456789abcdef".

More concretely, if you want to encrypt, say, a 16 digit number grouped into
4 groups of 4 using a `-` as a delimiter as in `0123-4567-8901-2345`, then you
would need a radix of at least 11, and you would need to translate the `-`
character to an `a` (as that is the value that follows `9`) prior to the
encryption. Conversely, you would need to translate an `a` to a `-` after
decryption.

This mapping of user inputs to alphabets defined by the radix is not performed
by the library and must be done prior to calling the encrypt and after calling
the decrypt functions.

A radix of up to 36 is supported, and the alphabet for a radix of 36 is
"0123456789abcdefghijklmnopqrstuvwxyz".

### Tweaks

Tweaks are very much like Initialization Vectors (IVs) in "traditional"
encryption algorithms. For FF1, the minimun and maximum allowed lengths of
the tweak may be specified by the user, and any tweak length between those
values may be used. For FF3-1, the size of the tweak is fixed at 7 bytes.

### Plain/ciphertext input lengths

For both FF1 and FF3-1, the minimum length is determined by the inequality:
- radix<sup>minlen</sup> >= 1000000

or:
- minlen >= 6 / log<sub>10</sub> radix

Thus, the minimum length is determined by the radix and is automatically
calculated from it.

For FF1, the maximum input length is
- 2<sup>32</sup>

For FF3-1, the maximum input length is
- 2 * log<sub>radix</sub> 2<sup>96</sup>

or:
- 192 / log<sub>2</sub> radix

## Examples

The [unit test code](lib/src/test/java/ubiqsecurity/fpe) provides the best
and simplest example of how to use the interfaces.

### FF1
```java
    /*
     * @key is a byte array whose length must be 16, 24, or 32
     * @twk is a byte array whose length must be between the minimum
     *      and maximum specified in the arguments to the constructor
     *
     * @radix and @PT are "user inputs"
     */
    String out;
    FF1 ctx;

    ctx = new FF1(key, twk, 0, 0, radix);

    out = ctx.encrypt(PT);
    out = ctx.decrypt(out);
```
### FF3-1
```java
    /*
     * @key is a byte array whose length must be 16, 24, or 32
     * @twk is a byte array whose length must be 7
     *
     * @radix and @PT are "user inputs"
     */
    String out;
    FF3_1 ctx;

    ctx = new FF3_1(key, twk, radix);

    out = ctx.encrypt(PT);
    out = ctx.decrypt(out);
```

[800-38g1]:https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38Gr1-draft.pdf
[ff1-examples]:https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/FF1samples.pdf
[ff3-cryptanalysis]:https://csrc.nist.gov/News/2017/Recent-Cryptanalysis-of-FF3


CHANGELOG.md
# Changelog

## 2.1.1 2023-10-16
* Bouncy Castle 1.76 JDK18ON upgrade

## 2.1.0 2023-09-05
* Change to compile for Java 1.8

## 2.0.0 2023-04-20
* Updated bouncy castle version and update to version numbers

## 0.1.4 2023-01-09
* Change algorithm to support arbitrary radix and charactersets

## 0.1.3 2022-03-03 
* Fix cipher algorithm issue and added additional unit tests

## 2022-02-14
* Fixed incorrect validation check

## 0.1.2 - 2021-09-29
* Added code to publish to sonatype / maven

## 0.1.1 - 2021-09-29
* Added support for radix and big integer bug fix 

## 0.1.0 - 2021-05-10
* Initial release



lib/src/test/java/ubiqfpe/FF1Test.java
package ubiqfpe;

import java.util.Arrays;

import org.junit.Test;
import static org.junit.Assert.*;

public class FF1Test
{
    private void test(final byte[] key, final byte[] twk,
                      final String PT, final String CT,
                      final int radix) {
        String out;
        FF1 ctx;

        assertEquals(PT.length(), CT.length());

        ctx = new FF1(key, twk, 0, 0, radix);

        out = ctx.encrypt(PT);
        assertEquals(CT, out);

        out = ctx.decrypt(CT);
        assertEquals(PT, out);
    }

    private final byte[] key = {
        (byte)0x2b, (byte)0x7e, (byte)0x15, (byte)0x16,
        (byte)0x28, (byte)0xae, (byte)0xd2, (byte)0xa6,
        (byte)0xab, (byte)0xf7, (byte)0x15, (byte)0x88,
        (byte)0x09, (byte)0xcf, (byte)0x4f, (byte)0x3c,
        (byte)0xef, (byte)0x43, (byte)0x59, (byte)0xd8,
        (byte)0xd5, (byte)0x80, (byte)0xaa, (byte)0x4f,
        (byte)0x7f, (byte)0x03, (byte)0x6d, (byte)0x6f,
        (byte)0x04, (byte)0xfc, (byte)0x6a, (byte)0x94,
    };

    private final byte[] twk1 = {};
    private final byte[] twk2 = {
        (byte)0x39, (byte)0x38, (byte)0x37, (byte)0x36,
        (byte)0x35, (byte)0x34, (byte)0x33, (byte)0x32,
        (byte)0x31, (byte)0x30,
    };
    private final byte[] twk3 = {
        (byte)0x37, (byte)0x37, (byte)0x37, (byte)0x37,
        (byte)0x70, (byte)0x71, (byte)0x72, (byte)0x73,
        (byte)0x37, (byte)0x37, (byte)0x37,
    };

    private final String[] PT = {
        "0123456789", "0123456789abcdefghi"
    };

    @Test
    public void nist1() {
        this.test(Arrays.copyOf(this.key, 16),
                  this.twk1,
                  PT[0], "2433477484", 10);
    }

    @Test
    public void nist2() {
        this.test(Arrays.copyOf(this.key, 16),
                  this.twk2,
                  PT[0], "6124200773", 10);
    }

    @Test
    public void nist3() {
        test(Arrays.copyOf(this.key, 16),
             this.twk3,
             PT[1], "a9tv40mll9kdu509eum", 36);
    }

    @Test
    public void nist4() {
        test(Arrays.copyOf(this.key, 24),
             this.twk1,
             PT[0], "2830668132", 10);
    }

    @Test
    public void nist5() {
        test(Arrays.copyOf(this.key, 24),
             this.twk2,
             PT[0], "2496655549", 10);
    }

    @Test
    public void nist6() {
        test(Arrays.copyOf(this.key, 24),
             this.twk3,
             PT[1], "xbj3kv35jrawxv32ysr", 36);
    }

    @Test
    public void nist7() {
        test(Arrays.copyOf(this.key, 32),
             this.twk1,
             PT[0], "6657667009", 10);
    }

    @Test
    public void nist8() {
        test(Arrays.copyOf(this.key, 32),
             this.twk2,
             PT[0], "1001623463", 10);
    }

    @Test
    public void nist9() {
        test(Arrays.copyOf(this.key, 32),
             this.twk3,
             PT[1], "xs8a0azh2avyalyzuwd", 36);
    }
}


lib/src/test/java/ubiqfpe/FFXTest.java
package ubiqfpe;

import org.junit.Test;
import static org.junit.Assert.*;

import java.util.Arrays;

import java.math.BigInteger;

public class FFXTest
{
    @Test
    public void str() {
        String s;

        s = FFX.str(5, 10, new BigInteger("12345", 10));
        assertEquals("12345", s);

        s = FFX.str(6, 10, new BigInteger("12345", 10));
        assertEquals("012345", s);

        assertThrows(RuntimeException.class, () -> {
                FFX.str(4, 10, new BigInteger("12345", 10));
            });
    }

    @Test
    public void rev() {
        String s;
        byte[] b;

        b = FFX.rev(new byte[]{ 1, 2, 3, 4 });
        assertArrayEquals(new byte[]{ 4, 3, 2, 1 }, b);

        b = FFX.rev(new byte[]{ 1, 2, 3, 4, 5 });
        assertArrayEquals(new byte[]{ 5, 4, 3, 2, 1 }, b);

        s = FFX.rev("abcd");
        assertEquals("dcba", s);

        s = FFX.rev("abcde");
        assertEquals("edcba", s);
    }
}


lib/src/test/java/ubiqfpe/FF3_1Test.java
package ubiqfpe;

import java.util.Arrays;

import org.junit.Test;
import static org.junit.Assert.*;

public class FF3_1Test
{
    private void test(final byte[] key, final byte[] twk,
                      final String PT, final String CT,
                      final int radix) {
        String out;
        FF3_1 ctx;

        assertEquals(PT.length(), CT.length());

        ctx = new FF3_1(key, twk, radix);

        out = ctx.encrypt(PT);
        assertEquals(CT, out);

        out = ctx.decrypt(CT);
        assertEquals(PT, out);
    }

    private final byte[] key = {
        (byte)0xef, (byte)0x43, (byte)0x59, (byte)0xd8,
        (byte)0xd5, (byte)0x80, (byte)0xaa, (byte)0x4f,
        (byte)0x7f, (byte)0x03, (byte)0x6d, (byte)0x6f,
        (byte)0x04, (byte)0xfc, (byte)0x6a, (byte)0x94,
        (byte)0x3b, (byte)0x80, (byte)0x6a, (byte)0xeb,
        (byte)0x63, (byte)0x08, (byte)0x27, (byte)0x1f,
        (byte)0x65, (byte)0xcf, (byte)0x33, (byte)0xc7,
        (byte)0x39, (byte)0x1b, (byte)0x27, (byte)0xf7,
    };

    private final byte[] twk1 = {
        (byte)0x00, (byte)0x00, (byte)0x00, (byte)0x00,
        (byte)0x00, (byte)0x00, (byte)0x00,
    };
    private final byte[] twk2 = {
        (byte)0x39, (byte)0x38, (byte)0x37, (byte)0x36,
        (byte)0x35, (byte)0x34, (byte)0x33,
    };
    private final byte[] twk3 = {
        (byte)0x37, (byte)0x37, (byte)0x37, (byte)0x37,
        (byte)0x70, (byte)0x71, (byte)0x72,
    };

    private final String[] PT = {
        "890121234567890000", "89012123456789abcde"
    };

    @Test
    public void ubiq1() {
        this.test(Arrays.copyOf(this.key, 16),
                  this.twk1,
                  PT[0], "075870132022772250", 10);
    }

    @Test
    public void ubiq2() {
        this.test(Arrays.copyOf(this.key, 16),
                  this.twk2,
                  PT[0], "251467746185412673", 10);
    }

    @Test
    public void ubiq3() {
        test(Arrays.copyOf(this.key, 16),
             this.twk3,
             PT[1], "dwb01mx9aa2lmi3hrfm", 36);
    }

    @Test
    public void ubiq4() {
        test(Arrays.copyOf(this.key, 24),
             this.twk1,
             PT[0], "327701863379108161", 10);
    }

    @Test
    public void ubiq5() {
        test(Arrays.copyOf(this.key, 24),
             this.twk2,
             PT[0], "738670454850774517", 10);
    }

    @Test
    public void ubiq6() {
        test(Arrays.copyOf(this.key, 24),
             this.twk3,
             PT[1], "o3a1og390b5uduvwyw5", 36);
    }

    @Test
    public void ubiq7() {
        test(Arrays.copyOf(this.key, 32),
             this.twk1,
             PT[0], "892299037726855422", 10);
    }

    @Test
    public void ubiq8() {
        test(Arrays.copyOf(this.key, 32),
             this.twk2,
             PT[0], "045013216693726967", 10);
    }

    @Test
    public void ubiq9() {
        test(Arrays.copyOf(this.key, 32),
             this.twk3,
             PT[1], "0sxaooj0jjj5qqfomh8", 36);
    }
}


lib/src/main/java/ubiqsecurity/fpe/Bn.java
package ubiqsecurity.fpe;

import java.io.IOException ;
import java.math.BigInteger;
 

/**
 * Algorithms to convert a numerical value in a given alphabet to a number
 */
 public class Bn {

    /**
     * Convert a numerical value in a given alphabet to a number.
     *
     * An alphabet consists of single-byte symbols in which each
     * symbol represents the numerical value associated with its
     * index/position in the alphabet. for example, consider the
     * alphabet: !@#$%^-*()
     * In this alphabet ! occupies index 0 and is therefore
     * assigned that value. @ = 1, # = 2, etc. Furthermore, the
     * alphabet contains 10 characters, so that becomes the radix
     * of the input. Using the alphabet above, an input of @$#
     * translates to a value of 132 (one hundred thirty-two,
     * decimal).
     *
     * If the alphabet above were instead: !@#$%^-*
     * The radix would be 8 and an input of @$# translates to a
     * value of 90 (ninety, decimal).
     *
     * @param str the numerical value to be converted
     * @param alpha alphabet consists of single-byte symbols
     *
     * @return the numerical value of the str pattern 
     * position found in the alphabet
     */
    public static BigInteger __bigint_set_str(final String str, final String alpha)
        {
            final int len = str.length();
            /*
             * the alphabet can be anything and doesn't have
             * to be in a recognized canonical order. the only
             * requirement is that every value in the list be
             * unique. checking that constraint is an expensive
             * undertaking, so it is assumed. as such, the radix
             * is simply the number of characters in the alphabet.
             */
            final int rad = alpha.length();
            if (rad <= 0) {
                throw new IllegalArgumentException("invalid argument, alphabet cannot be empty");
            }

            BigInteger m, a;
            int i;
            BigInteger x;
            
            /* represents the numerical value of str */
            x = BigInteger.valueOf(0);
            /*
             * multiplier used to multiply each digit
             * of the input into its correct position
             */
            m = BigInteger.valueOf(1);
            
            for (i = 0; i < len; i++) {
                final int pos;
                /*
                 * determine index/position in the alphabet.
                 * if the character is not present the input
                 * is not valid.
                 */
                pos = alpha.indexOf(str.charAt(len - 1 - i));
                if (pos < 0) {
                    throw new IllegalArgumentException("invalid argument, input character not found in alphabet");
                }
                /*
                 * multiply the digit into the correct position
                 * and add it to the result
                 */
                a = m.multiply(BigInteger.valueOf(pos));
                x = x.add(a);
                m = m.multiply(BigInteger.valueOf(rad));
            }
            return x;
        }
    
    
    /**
     * Inserts a character at a position in a String.
     *
     * Convenience function returns String with inserted char 
     * at an index position.
     *
     * @param str the original String
     * @param ch the character to insert
     * @param position the index position where to insert the ch
     *
     * @return    the new String containing the inserted ch 
     */    
    public static String insertChar(String str, char ch, int position) {
        StringBuilder sb = new StringBuilder(str);
        sb.insert(position, ch);
        return sb.toString();
    }
    
    
    /**
     * Gets the str pattern of the alphabet given the numeric value.
     *
     * @param alpha alphabet consists of single-byte symbols
     * @param x the numerical value of the str pattern
     *
     * @return the new String of the converted value 
     */    
    public static String __bigint_get_str(final String alpha, final BigInteger x) {
        final int rad = alpha.length();
        BigInteger quotient = x;
        String str = "";
        
        if (rad <= 0) {
            throw new IllegalArgumentException("invalid argument, alphabet cannot be empty");
        }
        
        /*
         * to convert the numerical value, repeatedly
         * divide (storing the resulted quotient and the remainder)
         * by the desired radix of the output.
         *
         * the remainder is the current digit; the result
         * of the division becomes the input to the next
         * iteration
         */
        while (quotient.compareTo(BigInteger.valueOf(0)) != 0) {
            int remainder;
            
            BigInteger result[] = quotient.divideAndRemainder(BigInteger.valueOf(rad));
            remainder = result[1].intValue();
            quotient = result[0];
            str = insertChar(str, alpha.charAt(remainder), 0);
        }
        
        if (str.length() == 0) {
            str = insertChar(str, alpha.charAt(0), 0);
        }
                
        return str;
    }

}


















lib/src/main/java/ubiqsecurity/fpe/FF1.java
package ubiqsecurity.fpe;

import java.math.BigInteger;
import java.util.Arrays;

/**
 * FF1 algorithm for format-preserving encryption
 */
public class FF1 extends FFX
{
    /**
     * Constructs a new context object for the FF1 algorithm.
     *
     * @param key     a byte array containing the key
     * @param twk     a byte array containing the "tweak" or iv. this value
     *                may not be null, and the number of bytes must be between
     *                the minimum and maximum allowed sizes
     * @param twkmin  the minimum number of bytes allowable for a tweak
     * @param twkmax  the maximum number of bytes allowable for a tweak or
     *                0 to indicate that there is no maximum
     * @param radix   the radix of the alphabet used for the plain and cipher
     *                text inputs/outputs
     */
    public FF1(final byte[] key, final byte[] twk,
               final long twkmin, final long twkmax,
               final int radix) {
        super(key, twk, (long)1 << 32, twkmin, twkmax, radix, FFX.DEFAULT_ALPHABET);
    }

    public FF1(final byte[] key, final byte[] twk,
      final long twkmin, final long twkmax,
      final int radix, final String alpha) {
      super(key, twk, (long)1 << 32, twkmin, twkmax, radix, alpha);
    }

    /*
     * The comments below reference the steps of the algorithm described here:
     *
     * https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38Gr1-draft.pdf
     */
    protected String cipher(final String X, byte[] twk, final boolean encrypt) {
        /* Step 1 */
        final int n = X.length();
        final int u = n / 2, v = n - u;

        /* Step 3, 4 */
        final int b = ((int)Math.ceil(
                           (Math.log(this.radix) / Math.log(2)) * v) + 7) / 8;
        final int d = 4 * ((b + 3) / 4) + 4;

        final int p = 16;
        final int r = ((d + 15) / 16) * 16;

        String A, B;
        byte[] PQ, R;
        int q;

        /* use default tweak if none is supplied */
        if (twk == null) {
            twk = this.twk;
        }

        /* check text and tweak lengths */
        if (n < this.txtmin || n > this.txtmax) {
            throw new IllegalArgumentException("invalid input length");
        } else if (twk.length < this.twkmin ||
                   (this.twkmax > 0 && twk.length > this.twkmax)) {
            throw new IllegalArgumentException("invalid tweak length");
        }

        /* the number of bytes in Q */
        q = ((twk.length + b + 1 + 15) / 16) * 16;

        /*
         * P and Q need to be adjacent in memory for the
         * purposes of encryption
         */
        PQ = new byte[p + q];
        R  = new byte[r];

        /* Step 2 */
        if (encrypt) {
            A = X.substring(0, u);
            B = X.substring(u);
        } else {
            B = X.substring(0, u);
            A = X.substring(u);
        }
        /* Step 5 */
        PQ[0]  = 1;
        PQ[1]  = 2;
        PQ[2]  = 1;
        PQ[3]  = (byte)(this.radix >> 16);
        PQ[4]  = (byte)(this.radix >>  8);
        PQ[5]  = (byte)(this.radix >>  0);
        PQ[6]  = 10;
        PQ[7]  = (byte)u;
        PQ[8]  = (byte)(n >> 24);
        PQ[9]  = (byte)(n >> 16);
        PQ[10] = (byte)(n >>  8);
        PQ[11] = (byte)(n >>  0);
        PQ[12] = (byte)(twk.length >> 24);
        PQ[13] = (byte)(twk.length >> 16);
        PQ[14] = (byte)(twk.length >>  8);
        PQ[15] = (byte)(twk.length >>  0);

        /* Step 6i, the static parts */
        System.arraycopy(twk, 0, PQ, p, twk.length);
        /* remainder of Q already initialized to 0 */

        for (int i = 0; i < 10; i++) {
          /* Step 6v */
            final int m = (((i + (encrypt ? 1 : 0)) % 2) == 1) ? u : v;

            BigInteger c, y;
            byte[] numb;

            /* Step 6i, the non-static parts */
            PQ[PQ.length - b - 1] = (byte)(encrypt ? i : (9 - i));

            /*
             * convert the numeral string B to an integer and
             * export that integer as a byte array into Q
             */
            c = FFX.number(B, this.radix, this.alpha); //new BigInteger(B, this.radix);
            numb = c.toByteArray();
            if (numb[0] == 0 && numb.length > 1) {
                /*
                 * Per the Java documentation, BigInteger.toByteArray always
                 * returns enough bytes to contain a sign bit. For the purposes
                 * of this function all numbers are unsigned; however, when the
                 * most-significant bit is set in a number, the Java library
                 * returns an extra most-significant byte that is set to 0.
                 * That byte must be removed for the cipher to work correctly.
                 */
                numb = Arrays.copyOfRange(numb, 1, numb.length);
            }
            if (b <= numb.length) {
                System.arraycopy(numb, 0, PQ, PQ.length - b, b);
            } else {
                /* pad on the left with zeros */
                Arrays.fill(PQ, PQ.length - b,
                            PQ.length - numb.length,
                            (byte)0);
                System.arraycopy(numb, 0,
                                 PQ, PQ.length - numb.length,
                                 numb.length);
            }

            /* Step 6ii */
            this.prf(R, 0, PQ, 0, PQ.length);

            /*
             * Step 6iii
             * if r is greater than 16, fill the subsequent blocks
             * with the result of ciph(R ^ 1), ciph(R ^ 2), ...
             */
            for (int j = 1; j < r / 16; j++) {
                final int l = j * 16;

                Arrays.fill(R, l, l + 12, (byte)0);
                R[l + 12] = (byte)(j >> 24);
                R[l + 13] = (byte)(j >> 16);
                R[l + 14] = (byte)(j >>  8);
                R[l + 15] = (byte)(j >>  0);

                FFX.xor(R, l, R, 0, R, l, 16);

                this.ciph(R, l, R, l);
            }

            /*
             * Step 6vi
             * calculate A +/- y mod radix**m
             * where y is the number formed by the first d bytes of R
             */
            y = new BigInteger(Arrays.copyOf(R, d));
            y = y.mod(BigInteger.ONE.shiftLeft(8 * d));

            c = FFX.number(A, this.radix, this.alpha); //new BigInteger(A, this.radix);
            if (encrypt) {
                c = c.add(y);
            } else {
                c = c.subtract(y);
            }

            c = c.mod(BigInteger.valueOf(this.radix).pow(m));

            /* Step 6viii */
            A = B;
            /* Step 6vii, 6ix */
            B = FFX.str(m, this.radix, this.alpha, c);
        }

        /* Step 7 */
        return encrypt ? (A + B) : (B + A);
    }
}


lib/src/main/java/ubiqsecurity/fpe/FFX.java
package ubiqsecurity.fpe;

import java.math.BigInteger;

import java.util.Arrays;

import javax.crypto.Cipher;

import org.bouncycastle.crypto.modes.CBCBlockCipher;
import org.bouncycastle.crypto.engines.AESEngine;
import org.bouncycastle.crypto.params.KeyParameter;

abstract class FFX
{

    public static final String DEFAULT_ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz";

    protected CBCBlockCipher cipher;
    protected int radix;
    protected long txtmin, txtmax;
    protected long twkmin, twkmax;
    protected byte[] twk;
    protected String alpha;

    protected FFX(final byte[] key, final byte[] twk,
                  final long txtmax,
                  final long twkmin, final long twkmax,
                  final int radix, final String alpha) {
        long txtmin;

        /* all 3 key sizes of AES are supported */
        switch (key.length) {
        case 16:
        case 24:
        case 32:
            break;
        default:
            throw new IllegalArgumentException("key size error");
        }

        /*
         * FF1 and FF3-1 support a radix up to 65536, but the
         * implementation becomes increasingly difficult and
         * less useful in practice after the limits below.
         */
        if (radix < 2 || radix > alpha.length()) {
            throw new IllegalArgumentException("invalid radix");
        }

        /*
         * for both ff1 and ff3-1: radix**minlen >= 1000000
         *
         * therefore:
         *   minlen = ceil(log_radix(1000000))
         *          = ceil(log_10(1000000) / log_10(radix))
         *          = ceil(6 / log_10(radix))
         */
        txtmin = (int)Math.ceil(6.0 / Math.log10(radix));
        if (txtmin < 2 || txtmin > txtmax) {
            throw new RuntimeException("minimum text length out of range");
        }

        /* the default tweak must be specified */
        if (twk == null) {
            throw new NullPointerException("invalid tweak");
        }
        /* check tweak lengths */
        if (twkmin > twkmax ||
            twk.length < twkmin ||
            (twkmax > 0 && twk.length > twkmax)) {
            throw new IllegalArgumentException("invalid tweak length");
        }

        /*
         * the underlying cipher for FF1 and FF3-1 is AES in CBC mode.
         * by not specifying the IV, the IV is set to 0's which is
         * what is called for in these algorithms
         */
        this.cipher = new CBCBlockCipher(new AESEngine());
        this.cipher.init(true, new KeyParameter(key));

        this.radix = radix;
        this.alpha = alpha;

        this.txtmin = txtmin;
        this.txtmax = txtmax;

        this.twkmin = twkmin;
        this.twkmax = twkmax;

        this.twk = Arrays.copyOf(twk, twk.length);
    }

    abstract protected String cipher(
        final String X, byte [] twk, final boolean encrypt);

    /*
     * perform an aes-cbc encryption (with an IV of 0) of @src, storing
     * the last block of output into @dst. The number of bytes in @src
     * must be a multiple of 16. @dst and @src may point to the same
     * location but may not overlap, otherwise. @dst must point to a
     * location at least 16 bytes long
     */
    protected void prf(byte[] dst, final int doff,
                       final byte[] src, final int soff, final int len) {
        final int blksz = this.cipher.getBlockSize();

        if ((src.length - soff) % blksz != 0) {
            throw new IllegalArgumentException("invalid source length");
        }

        // Some time, we want to run through process block for the entire src
        // sometimes just one block of the src, regardless of the length.
        // In cases where only one block needs to be processed, len would be
        // block size and will terminate the look.  In othercases, len will
        // be the size of the src but len - soff will terminate that.  however
        // cannot easily combine both checks into a single math equation.
        for (int i = 0; i < len && i < src.length - soff; i += blksz) {
            this.cipher.processBlock(src, soff + i, dst, doff);
        }
        this.cipher.reset();
    }

    /*
     * perform an aes-ecb encryption of @src. @src and @dst must each be
     * 16 bytes long, starting from the respective offsets. @src and @dst
     * may point to the same location or otherwise overlap
     */
    protected void ciph(byte[] dst, final int doff,
                        final byte[] src, final int soff) {
        this.prf(dst, doff, src, soff, 16);
    }

    /*
     * a convenience version of the ciph function that returns its
     * output as a separate byte array
     */
    protected byte[] ciph(final byte[] src) {
        byte[] dst = new byte[this.cipher.getBlockSize()];
        ciph(dst, 0, src, 0);
        return dst;
    }

    /*
     * reverse the bytes in a byte array. @dst and @src may point
     * to the same location but may not otherwise overlap
     */
    public static void rev(byte[] dst, final byte[] src) {
        int i;

        for (i = 0; i < src.length / 2; i++) {
            final byte t = src[i];
            dst[i] = src[src.length - i - 1];
            dst[src.length - i - 1] = t;
        }

        if (src.length % 2 == 1) {
            dst[i] = src[i];
        }
    }

    /*
     * convenience function that returns the reversed sequence
     * of bytes as a new byte array
     */
    public static byte[] rev(final byte[] src) {
        byte[] dst = new byte[src.length];
        rev(dst, src);
        return dst;
    }

    /*
     * reverse the characters in a string
     */
    public static String rev(final String str) {
        StringBuilder sb =  new StringBuilder(str);
        return sb.reverse().toString();
    }

    /*
     * Perform an exclusive-or of the corresponding bytes
     * in two byte arrays
     */
    public static void xor(byte[] d, final int doff,
                           final byte[] s1, final int s1off,
                           final byte[] s2, final int s2off,
                           final int len) {
        for (int i = 0; i < len; i++) {
            d[doff + i] = (byte)(s1[s1off + i] ^ s2[s2off + i]);
        }
    }

    /*
     * convert a big integer to a string under the radix @r with
     * length @m. If the string is longer than @m, the function fails.
     * if the string is shorter that @m, it is zero-padded to the left
     */
    public static String str(final int m, final int r, final BigInteger i) {
      return str(m, r, FFX.DEFAULT_ALPHABET, i);
    }

    public static String str(final int m, final String alpha, final BigInteger i) {
      // String s = i.toString(r);
      return str(m, alpha.length(), alpha, i);
  }

  public static String str(final int m, final int radix, final String alpha, final BigInteger i) {
    // String s = i.toString(r);
    StringBuilder sb = new StringBuilder();
    BigInteger bi_radix = BigInteger.valueOf(radix);
    BigInteger cvt = i;

    while (cvt.compareTo(BigInteger.ZERO) > 0) {
        sb.insert(0, alpha.charAt(cvt.mod(bi_radix).intValue()));
        cvt = cvt.divide(bi_radix);
    }
    if (sb.length() > m) 
    {
      throw new RuntimeException(String.format("Unable to convert biginteger into %d characters",m));
    }
    while (sb.length() < m)
    {
        sb.insert(0, alpha.charAt(0));
    }
    return sb.toString();
  }

  public static BigInteger number(final String num, final int radix) {
    return number(num, radix, FFX.DEFAULT_ALPHABET);
  }

  public static BigInteger number(final String num, final String alpha) {
    return number(num, alpha.length(), alpha);
  }

  public static BigInteger number(String num, int radix, String alpha) {
    BigInteger ret = BigInteger.valueOf(0);
    int idx = num.length() - 1;
    BigInteger digit = BigInteger.valueOf(1);
    BigInteger bi_radix = BigInteger.valueOf(radix);
    while (idx >= 0) {
      ret = ret.add(digit.multiply(BigInteger.valueOf(alpha.indexOf(num.charAt(idx)))));
      idx--;
      digit = digit.multiply(bi_radix);
    }
    return ret;
  }

    /**
     * Encrypt a string, returning a cipher text using the same alphabet.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   the plain text to be encrypted
     * @param twk the tweak used to perturb the encryption
     *
     * @return    the encryption of the plain text, the cipher text
     */
    public String encrypt(String X, byte[] twk) {
        return this.cipher(X, twk, true);
    }

    /**
     * Encrypt a string, returning a cipher text using the same alphabet.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   The plain text to be encrypted
     *
     * @return    the encryption of the plain text, the cipher text
     */
    public String encrypt(String X) {
        return this.encrypt(X, null);
    }

    /**
     * Decrypt a string, returning the plain text.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   the cipher text to be decrypted
     * @param twk the tweak used to perturb the encryption
     *
     * @return    the decryption of the cipher text, the plain text
     */
    public String decrypt(String X, byte[] twk) {
        return this.cipher(X, twk, false);
    }

    /**
     * Decrypt a string, returning the plain text.
     *
     * The key, tweak parameters, and radix were all already set
     * by the initialization of the FF3_1 object.
     *
     * @param X   the cipher text to be decrypted
     *
     * @return    the decryption of the cipher text, the plain text
     */
    public String decrypt(String X) {
        return this.decrypt(X, null);
    }
}
